### ISSUE TITLE
Deal with ICmpInitializable.OnInit

### ISSUE THREAD

**Issue by AdamsLair** [AdamsLair/duality#136]
- Having to check for the context parameter is counter-intuitive and a newbie pitfall.
- In 99% of cases, what the user really wants is just an OnActivate / OnDeactivate
- Consider splitting ICmpInitializable into multiple interfaces, where ICmpInitializable will just deal with activation states.
  - Might keep the OnInit / OnShutdown naming to not throw everyone off the track.
- Also, evaluate which of those are really needed. Get rid of all others.

**Comment by AdamsLair** [comment]
Right now, there are four `InitContext` values and three `ShutdownContext` values:

**Init**:
- **`Saved`**: A saving process has just finished.
  - Completely unused.
  - Could be easily removed.
- **`Loaded`**: The Component has been fully loaded.
  - Used in AnimSpriteRenderer, TextRenderer and Transform for a first initialization when loading the Scene.
  - Can't be removed.
- **`Activate`**: The Component is being activated. This can be the result of activating it, activating its GameObject, adding itsself or its GameObject to the current Scene or entering a Scene in which this Component is registered.
  - Used everywhere.
  - This is the core feature here.
- **`AddToGameObject`**: The Component has just been added to a GameObject.
  - Used in Transform for initializing event handlers, and in RigidBody for adding the default circle in the editor.
  - Can't be removed.

**Shutdown**:
- **`Saving`**: A saving process is about to start.
  - Used in RigidBody for cleanup and in VisualLogRenderer to remove the object and prevent it from being saved persistently in the Scene.
  - Can't be removed as is, but could be when finding a proper way to express the two use cases otherwise.
- **`Deactivate`**: The Component has been deactivated. This can be the result of deactivating it, deactivating its GameObject, removing itsself or its GameObject from the current Scene or leaving a Scene in which this Component is registered.
  - Used everywhere.
  - This is the core feature here.
- **`RemovingFromGameObject`**: The Component is being removed from its GameObject.
  - Used in Transform for shutting down event handlers.
  - Can't be removed.

Summing this up, while singular contexts could be removed, no category of contexts could be removed as a whole, leaving the removal of each single context useless for the course of a general cleanup.

As far as regrouping and API changes go, the following would be one way to do it:

- **ICmpActivatable:** Activate / Deactivate
- **ICmp"ObjectAdjustable" (?):** AddToGameObject / RemoveFromGameObject
- **ICmp"SerializeListener" (?):**Loaded / Saving / Saved

Investigate.

**Comment by AdamsLair** [comment]
There's no community consensus on how to deal with this yet: [Forum Thread](http://forum.adamslair.net/viewtopic.php?f=4&t=547). It also isn't perceived as a critical change to do before v2.0 - consider moving it off the release schedule.

Also consider implementing the split interfaces method as mentioned above, since it appears to have the advantage of splitting required knowledge into smaller chunks.

**Comment by JAOT** [comment]
Why not adding the different methods when implementing ICmpInitializable? Yes, the class would become huge, but it could become more clear when using.
Or adding another interface for saving/loading for cleanup.

**Comment by ilexp** [comment]
> Why not adding the different methods when implementing ICmpInitializable? Yes, the class would become huge, but it could become more clear when using.

Because every user would have to implement all the methods in every single Component that requires an activation handler. It's just not very user friendly.

> Or adding another interface for saving/loading for cleanup.

Yep, that's the current idea as detailed in the above posts: Instead of having one interface for all initialization work, have multiple distinct interfaces for different init scenarios.

**Comment by RabbitB** [comment]
I had an idea about a possible way to go about solving this. I originally made a post in the forum replying to you @ilexp, but it's pretty long and I have some extra stuff to say about it, so I'm not going to reiterate everything here, and instead I'm just going to link it: [Maybe use functional replacement for ICmpX interfaces?](http://forum.adamslair.net/viewtopic.php?f=4&t=868&p=5243#p5243)

In short, go the functional route, where instead of the scene asking components (by looking at interfaces) whether they support an action, we have the components explicitly tell the scene what they support and give it the function to call for that action.

I can think of a number of benefits by going this route.
- No need to implement extraneous interfaces or methods.
- No need for the scene to spend time notifying components that don't care about an action, because the interface actually represents multiple actions (looking at you ICmpInitializable).
- Because we're just passing around functions, we can change the implementation of an action on a per instance basis. We have all, at some time or another, written huge monstrous update methods with tons of branches based on component, scene or game state. No need anymore; write smaller, more efficient and focused update methods, and select the correct one to declare to the scene. Even more useful if the scene has a `SwapAction` method that allows us to swap the function we declared on the fly (most likely would want to seriously limit how often you swap for performance considerations).
- A bit of a give/take situation, but this setup would require keeping lookup tables for each action, with the component as a key and its function for that action as a value. The upside is that you already track every component by the actions it has declared, so running `OnUpdate` is as simple as iterating through the `OnUpdate` lookup.

**Comment by ilexp** [comment]
It's an interesting idea, but my first impression is that it will do more harm than good overall. Here's why:

> - No need to implement extraneous interfaces or methods.
> - No need for the scene to spend time notifying components that don't care about an action, because the interface actually represents multiple actions (looking at you ICmpInitializable).

The way I see interfaces in this case is as a design contract: You are specifying that a certain class implements a certain API and you're forced to stick to that. In the case of Duality `ICmp` interface, you're specifying that your component is interested in, or can do certain things, while simultaneously providing the API to handle it accordingly. When looking at a Component, you can immediately see what core behaviors it exhibits and what it might react to, simply by looking at the interfaces it implements, which I think is really useful.

The case where the issues arise is in fact `ICmpInitializable`, because it's really just badly designed as a sort of "aggregate" interface. This needs to change, but it can be changed without moving away from the interface approach entirely.

> - Because we're just passing around functions, we can change the implementation of an action on a per instance basis. We have all, at some time or another, written huge monstrous update methods with tons of branches based on component, scene or game state. No need anymore; write smaller, more efficient and focused update methods, and select the correct one to declare to the scene. Even more useful if the scene has a SwapAction method that allows us to swap the function we declared on the fly (most likely would want to seriously limit how often you swap for performance considerations).

What you gain here is flexibility, but at the cost of maintainability. As soon as more than one developer is working on something, or even as soon as code is shared between developers, being able to do stuff like switching out notifier methods dynamically increases complexity a lot, both for reasoning about code and debugging it.

Right now, if you want to know what a component does when it's updated, you just have to do at `OnUpdate`, that's it. It will tell you all you need to know. As soon as you are able to specify a different notifier, the code you're interested in could be _anywhere_, and even worse, it could switch at any time and differ between object instances. And I haven't even started to think about the implications of this when it comes to a modular, plugin-based and in a lot of places declarative framework architecture that Duality has.

Wrapping it up:

- Do people want to be able to do this kind of thing? I don't count myself among the, but maybe.
- Do people want to debug this kind of thing and have to deal with the repercussions it has because everyone else is using it too? Pretty sure the answer is no.

In any case, thanks for contributing to the discussion with your idea!

**Comment by JAOT** [comment]
Again I post, but nothing as high level as the previous postings to show a way to go about this: [XNA Example](https://msdn.microsoft.com/en-us/library/bb203893.aspx#ID4EBF).

I believe a lot of newbie mistakes could be avoided by setting in stone the base foundations of a class, wich could make some Interfaces unnecessary, even if the available methods were to be completely empty.

Duality could do just that. Spawn a class with the necessary methods to get the ball running, with more options available through Interfaces.

**Comment by RabbitB** [comment]
>@ilexp In any case, even though I may be skeptical about this particular one, thanks for contributing to the discussion with your idea!

Always glad to; when designing an API, proper solutions to these issues are rarely found without suggestions that look at the issue from every angle, even if they aren't ideal. It's not so much the suggestion itself that's important, but that it inspires us to look at an issue in ways we haven't before.

**Comment by Barsonax** [comment]
What about keeping it simple and simply adding 1 interface for the common use case while keeping the old ICmpInitializable interface for ppl who need the other contexts. There are a couple advantages:
-Does not break any older code so could be implemented right away, no need to wait for v3.
-Simple to implement.
-Covers almost all cases since this is the common case.

A disadvantage would be redundancy though as ICmpInitializable does provide the same functionality.

**Comment by ilexp** [comment]
Addressed this in the new [feature/3.0/initializable-refactoring](https://github.com/AdamsLair/duality/tree/feature/3.0/initializable-refactoring) branch.

- `ICmpInitializable` was refactored to only have two context-less `OnActivate` / `OnDeactivate` methods.
- `ICmpSerializeListener` provides `OnLoaded`, `OnSaved` and `OnSaving` methods.
- `ICmpAttachmentListener` provides `OnAddToGameObject` and `OnRemoveFromGameObject` methods.
- `InitContext` and `ShutdownContext` enums were removed.
- All related code was updated to use the new API.
- Documented all related interfaces and interface methods more thoroughly.

Comparing the old vs. the new approach from the user side, before:
```csharp
void ICmpInitializable.OnInit(InitContext context)
{
    if (context == InitContext.Activate)
    {
        // Do some cleanup before updating again
        this.RemoveDisposedJoints();
        // Initialize the backing Farseer objects upon activation
        this.Initialize();
    }
}
void ICmpInitializable.OnShutdown(ShutdownContext context)
{
    if (context == InitContext.Deactivate)
    {
        this.Shutdown();
    }
}
```
After:
```csharp
void ICmpInitializable.OnActivate()
{
    // Do some cleanup before updating again
    this.RemoveDisposedJoints();
    // Initialize the backing Farseer objects upon activation
    this.Initialize();
}
void ICmpInitializable.OnDeactivate()
{
    this.Shutdown();
}
```
Looks much more straight-forward to me, especially for new users. Ready to merge back to `develop-3.0`.

**Comment by ilexp** [comment]
Merged. Closing this.
